pkg/deploy/internal/segment/segment_test.go

```diff
@@ -0,0 +1,348 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package segment_test
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/api"
+	"github.com/dynatrace/dynatrace-configuration-as-code-core/clients/segments"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/coordinate"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/entities"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/template"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/deploy/internal/segment"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"testing"
+)
+
+type testClient struct {
+	upsertStub func() (segments.Response, error)
+	getStub    func() (segments.Response, error)
+	getAllStub func() ([]segments.Response, error)
+}
+
+func (tc *testClient) Upsert(_ context.Context, _ string, _ []byte) (segments.Response, error) {
+	return tc.upsertStub()
+}
+
+func (tc *testClient) Get(_ context.Context, _ string) (segments.Response, error) {
+	return tc.getStub()
+}
+
+func (tc *testClient) GetAll(_ context.Context) ([]segments.Response, error) {
+	return tc.getAllStub()
+}
+
+func TestDeploy(t *testing.T) {
+	testCoordinate := coordinate.Coordinate{
+		Project:  "my-project",
+		Type:     "segment",
+		ConfigId: "my-config-id",
+	}
+	tests := []struct {
+		name           string
+		inputConfig    config.Config
+		upsertStub     func() (segments.Response, error)
+		getStub        func() (segments.Response, error)
+		getAllStub     func() ([]segments.Response, error)
+		expected       entities.ResolvedEntity
+		expectErr      bool
+		expectedErrMsg string
+	}{
+		{
+			name: "deploy with objectOriginId - success PUT",
+			inputConfig: config.Config{
+				Template:       template.NewInMemoryTemplate("path/file.json", "{}"),
+				Coordinate:     testCoordinate,
+				OriginObjectId: "my-object-id",
+				Type:           config.Segment{},
+				Parameters:     config.Parameters{},
+				Skip:           false,
+			},
+			upsertStub: func() (segments.Response, error) {
+				return segments.Response{
+					StatusCode: http.StatusOK,
+				}, nil
+			},
+			getStub: func() (segments.Response, error) {
+				return segments.Response{
+					StatusCode: http.StatusOK,
+					Data: marshal(map[string]any{
+						"uid":         "JMhNaJ0Zbf9",
+						"name":        "test-segment-post-match",
+						"description": "post - update from monaco - change - 2",
+						"isPublic":    false,
+						"owner":       "79a4c92e-379b-4cd7-96a3-78a601b6a69b",
+						"externalId":  "project_segment:segment:some-id",
+					}),
+				}, nil
+			},
+			getAllStub: func() ([]segments.Response, error) {
+				panic("should not be called")
```

Comment:
Any chance we can inject `testing.T` here and call `t.Fatal()`? I see that it's not trivial, but I'd prefer it to panicing 🤔

Reply:
I'll have a look for a more elegant solution 👍